### PR TITLE
[ESIMD] Add 'inline' to function decls to fix ODR problems.

### PR DESCRIPTION
**Pull request by kbobrovs**
Fixes "multiple definition" linker errors when multiple sources using ESIMD are linked together.

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/INTEL/esimd/detail/esimd_memory_intrin.hpp
Comment (kychendev) [inline sycl/include/CL/sycl/INTEL/esimd/detail/esimd_memory_intrin.hpp:469]: With this change, for non-sycl-device build, compiler FE will see one SYCL_EXTERNAL declaration and one inline definition for the same API, while the existing code made it consistent. What is the exact FE rule here, or compiler doesn't have diagnostic for this? Should similar change be made in other files such as esimd_math_intrin.hpp?

Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/detail/esimd_memory_intrin.hpp:469]: `inline` will be seen only by the host compiler, which is really the intent.
Device compiler will continue to see SYCL_EXTERNAL w/o definition

Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/detail/esimd_memory_intrin.hpp:469]: > Should similar change be made in other files such as esimd_math_intrin.hpp?

Yes. This is achieved by including `inline` into `ESIMD_INLINE` used in other files.
Also, SYCL_EXTERNAL is defined to empty string for the host compiler, so `void foo` followed by `inline void foo` should be OK.

Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/detail/esimd_memory_intrin.hpp:469]: Had a sync with @kychendev - esimd_math_intrin.hpp should be really fixed, my omission. Will do.

**Comment by kbobrovs** [comment]
It turns out all the intrinsic functions in `esimd_math_intrin.hpp` are templated. C++ ODR allows to have multiple definitions for template functions in a program as long as they are not specializations (always full for functions) and appear in different 
translation units (+ some more conditions, which are all satisfied in our case). See https://en.cppreference.com/w/cpp/language/definition

> One Definition Rule
> ...
> There can be more than one definition in a program of each of the following: class type, enumeration type, inline function, inline variable (since C++17), templated entity (template or member of template, but not full template specialization), as long as all of the following is true:

So there is no need to fix `esimd_math_intrin.hpp`, even though due to a different reason than I provided before.
@kychendev - please approve if you are OK with this resolution.

**Comment by kbobrovs** [comment]
> So there is no need to fix esimd_math_intrin.hpp

actually, some clean-up still would be good - `SYCL_EXTERNAL` is used for host-only definitions, which does not make sense, albeit harmless. Plus the test has an error, which prevents from correct execution - it is still good for intended purpose, but I'll fix it anyway.

**Comment by kychendev** [comment]
@kbobrovs, looks good but there is test error related to unsupported sycl target. Please double check.

**Comment by kbobrovs** [comment]
@kychendev, all checks now pass

**Review by kychendev — APPROVED** [review]

**Comment by kbobrovs** [comment]
@bader, please take a look.

**Comment by bader** [comment]
> @bader, please take a look.

@kbobrovs, I'm familiar with this code very well, so I think your approval as a code owner should be enough to merge this patch.

File: sycl/test/esimd/odr.cpp
Comment (bader) [inline sycl/test/esimd/odr.cpp:60]: Minor style comment.
```suggestion
  if (std::memcmp(gold, data, sizeof(int) * VLEN)) {
    std::cout << "FAILED\n";
    for (int i = 0; i < VLEN; i++)
      std::cout << " " << data[i];
    std::cout << "\n";
    return 1;
  }
  
  std::cout << "Passed\n";
  return 0;

```

**Review by bader — APPROVED** [review]

**Comment by kbobrovs** [comment]
> > @bader, please take a look.
> 
> @kbobrovs, I'm familiar with this code very well, so I think your approval as a code owner should be enough to merge this patch.

@bader, right, but in this case I'm also an author. But if you are OK with this, I will merge myself then. Thanks.

**Comment by bader** [comment]
> But if you are OK with this, I will merge myself then. Thanks.

I suggest adding one more person as a code owner for ESIMD, who is familiar with the code and can do code review for your changes.
I'm okay, but it doesn't count. :-)

**Comment by kbobrovs** [comment]
> I suggest adding one more person as a code owner for ESIMD

right, good idea. I created https://github.com/intel/llvm/pull/3114

**Comment by kbobrovs** [comment]
@DenisBakhvalov - could you please review/approve? @kychendev already approved (approval lost due to test re-formatting)

**Review by DenisBakhvalov — APPROVED** [review]

**Review by kychendev — APPROVED** [review]

**Pull request merged**
